docs/project/milestones.md

```diff
@@ -0,0 +1,305 @@
+    C++ language features except for coroutines, is documented, cohesive, and
+    understandable by evaluators without placeholders.
+-   Evaluators can build and run tests of most C++ interoperability, including
+    both real-world C++ code that can be built with the latest release of Clang
+    and test Carbon code.
+    -   Gaps from the design need to be ones that don't undermine evaluation
+        confidence.
+-   Evaluators can effectively stress-test build speed and scaling with Carbon.
+-   Evaluators can build some key benchmarks that include C++ interoperation in
+    the critical path and get representative performance results.
+    -   This can in turn be a still smaller subset of all aspects of C++
+        interoperability based around what impacts interesting benchmarks.
+
+### Language features
+
+These are focused on the core _necessary_ features for us to reach a successful
+0.1 language that can address our goals. Some of these features are required
+directly by the above goals, others are required due to dependencies or
+interactions with the directly required features. However, we don't try to cover
+all of the features in full granularity here. There will be many minor
+components that are necessary for these to hold together but are not directly
+addressed. In general, unless something is explicitly described as partial or
+having exceptions, everything covered by that entry should be expected in the
+0.1 language.
+
+Another important point is that this doesn't commit Carbon to any _particular_
+design for any of these bullet points. Instead, it just means that the Carbon
+design must have _something_ that addresses each of these bullet points. That
+might be to add the named design to Carbon, but it equally might be a clear
+statement that Carbon will _not_ include this design but use some other language
+features to address its use cases and when rewriting C++ code using that feature
+into Carbon.
+
+#### Code organization and structuring
+
+-   Packages
+-   Libraries
+-   Implementation files
+-   Importing
+-   Namespaces
+
+#### Type system
+
+-   User-defined types
+    -   C++ interop: importing C++ types into Carbon, exporting Carbon types
+        into C++
+    -   Single inheritance
+        -   Virtual dispatch
+        -   C++ interop:
+            -   Bi-directional inheritance between C++ and Carbon
+            -   Type hierarchy roots in both C++ and Carbon
+            -   Mappings of inheritance features: abstract, final, virtual
+    -   Operator overloading
+        -   C++ interop:
+            -   Synthesizing Carbon overloads for imported C++ types
+            -   Exporting Carbon overloads into C++
+    -   Sum types (discriminated unions)
+    -   Unions (un-discriminated)
+        -   C++ interop: mapping to and from C++ unions.
+-   Generics
+    -   Both generic functions and types
+    -   Checked generics
+        -   Definition-checked variadics
```

Comment:
If you're looking to punt something, this could be punted. Memory safety, IMO, is much more important than checked generics and variadics right now.

Reply:
FWIW, I agree that memory safety is going to be a *big* motivation.

The reason I've pushed to prioritize checked generics is because they form the foundation of the language -- we end up building most other language features in some part on top of the checked generics system. And without a reasonably complete design, it is very hard to know that the infrastructure there will continue to hold together. As just one example that impacts memory safety, if we end up needing a lifetime annotation system like Rust, we'll likely need a checked generics system to allow being generic over lifetimes without causing needless template instantiations.

Whether we need definition-checked *variadics* is I think a more interesting question. There I think the interop with C++ variadics is a larger part of the motivation. We could probably get by with a purely template system there, but I think we would end up ripping it out and replacing it later to get it to play well with things like lifetimes. Ultimately, I'd rather not put a placeholder into something that ends up being this fundamental, and I think we're close enough with checked variadics that we should really finish it off.